Gcp terraform modules docs (#1424)
pingcap/docs: v3.0/tidb-in-kubernetes/deploy/gcp-gke.md

{{< copyable "shell-regular" >}}

```bash
gcloud services enable cloudresourcemanager.googleapis.com \
    cloudbilling.googleapis.com cloud services enable iam.googleapis.com \
    compute.googleapis.com container.googleapis.com
```

### Configure Terraform

The terraform script expects three variables to be provided by the user. You can let Terraform prompt you for them, or define them in a `.tfvars` file of your choice. The three variables are:

> * `GCP_CREDENTIALS_PATH`: Path to a valid GCP credentials file.

Author: The GCP doc page that is linked for the service account key creation gives some details about the content. Is there anything that should be included in our docs that isn't listed? Specifically anything that might confuse an inexperienced user, potentially.

> * `GCP_CREDENTIALS_PATH`: Path to a valid GCP credentials file.

Reviewer: Not particularly. I think we are good here.

>     - It is recommended to have a separate service account to be used by Terraform. See [this page](https://cloud.google.com/iam/docs/creating-managing-service-accounts) for more information. The `Project Editor` role will grant sufficient permissions for provisioning the necessary infrastructure resources.

Reviewer 2: It is already implied that Project Editor is sufficient permissions to do anything.

```suggestion
    - It is recommended to have a separate service account to be used by Terraform. See [this page](https://cloud.google.com/iam/docs/creating-managing-service-accounts) for more information. `./create-service-account.sh` will create such a service account with minimal permissions.
```

>     - See [this page](https://cloud.google.com/iam/docs/creating-managing-service-account-keys) for information on creating service account keys. The steps below detail how to do this using a script provided in the `deploy/gcp` directory, alternatively if creating the service account and key yourself, choose `JSON` key type during creation. The downloaded `JSON` file that contains the private key is the credentials file you need.

Author: Specifically it's the block of text after
> Below we will show how to set these variables:

>     - See [this page](https://cloud.google.com/iam/docs/creating-managing-service-account-keys) for information on creating service account keys. The steps below detail how to do this using a script provided in the `deploy/gcp` directory, alternatively if creating the service account and key yourself, choose `JSON` key type during creation. The downloaded `JSON` file that contains the private key is the credentials file you need.

Reviewer: ```suggestion
    - See [this page](https://cloud.google.com/iam/docs/creating-managing-service-account-keys) for information on creating service account keys. The steps in the script below detail how to do this using a script provided in the `deploy/gcp` directory, alternatively if creating the service account and key yourself, choose `JSON` key type during creation. The downloaded `JSON` file that contains the private key is the credentials file you need.
```

* `GCP_REGION`: The region to create the resources in, for example: `us-west1`.
* `GCP_PROJECT`: The GCP project in which everything will be created.

> Below we will show how to set these variables:

Reviewer: ```suggestion
The script below shows how to configure Terraform with these variables:
```

> 

Reviewer 2: ^ This gcloud command takes multiple arguments, so this can be compressed into one line (maybe use 2 or 3 for better readability?)

### Configure Terraform

The terraform script expects three variables to be provided by the user. You can let Terraform prompt you for them, or define them in a `.tfvars` file of your choice. The three variables are:

* `GCP_CREDENTIALS_PATH`: Path to a valid GCP credentials file.

>     - It is recommended to create a new service account to be used by Terraform. See [this page](https://cloud.google.com/iam/docs/creating-managing-service-accounts) to create a service account and grant `Project Editor` role to it.

Reviewer 2: Maybe we should change this now that we have a script that automatically does this? Right now the user might start doing this here instead of just running the ./create-service-account.sh script below.

Author: I guess it depends what level of terraform  and GCP experience we're targeting. The steps to populate them come immediately afterwards, so a totally new user can follow those, but a more experienced user can see the three variables and fill them in on their own, if they choose. Open to suggestions on this.

Author: @gregwebs let me know what you think about the updated section

./create-service-account.sh
```

Terraform will automatically load and populate variables from files matching `terraform.tfvars` or `*.auto.tfvars`. For more information please see the [Terraform documentation](https://learn.hashicorp.com/terraform/getting-started/variables.html).

> The previous steps will populate `terraform.tfvars` with `GCP_REGION` and `GCP_PROJECT`, and `credentials.auto.tfvars`

Author: The previous steps here refers to this block

```bash
# Replace the region with your GCP region and your GCP project name.
echo GCP_REGION=us-west1 >> terraform.tfvars
# First make sure you are connected to the correct project. gcloud config set project $PROJECT
echo "GCP_PROJECT=$(gcloud config get-value project)" >> terraform.tfvars
# Create a service account for terraform with restricted permissions and set the credentails path
./create-service-account.sh
```

> The previous steps will populate `terraform.tfvars` with `GCP_REGION` and `GCP_PROJECT`, and `credentials.auto.tfvars`

Reviewer: Oh I thought there were some steps before that...

> The previous steps will populate `terraform.tfvars` with `GCP_REGION` and `GCP_PROJECT`, and `credentials.auto.tfvars`

Reviewer: ```suggestion
The steps in the above script will populate `terraform.tfvars` with `GCP_REGION` and `GCP_PROJECT`, and `credentials.auto.tfvars`
```

with `GCP_CREDENTIALS_PATH`.

## Deploy

> You need to decide on instance types. If you just want to get a feel for a TiDB deployment and lower your cost, you can use the small settings.

Reviewer: If you are fine with this edit, please also add the bullet symbol "-"  before the next method "If you want to benchmark a production deployment..."

1 row in set (0.001 sec)
```

## Manage multiple TiDB clusters

An instance of a `tidb-cluster` module corresponds to a TiDB cluster in the GKE cluster. If you want to add a new TiDB cluster,
you can edit the `tidbclusters.tf` file and add a new instance of the `tidb-cluster` module. For example:

```hcl
module "example-tidb-cluster" {
  providers = {
    helm = "helm.gke"
  }
  source                     = "../modules/gcp/tidb-cluster"
  gcp_project                = "gcp-project-name"
  gke_cluster_location       = "us-west1"
  gke_cluster_name           = "gke-cluster-name"
  cluster_name               = "example-tidb-cluster"
  cluster_version            = "v3.0.1"
  kubeconfig_path            = module.tidb-operator.kubeconfig_path
  tidb_cluster_chart_version = "v1.0.0"
  pd_instance_type           = "n1-standard-1"
  tikv_instance_type         = "n1-standard-4"
  tidb_instance_type         = "n1-standard-2"
  monitor_instance_type      = "n1-standard-1"
  pd_node_count              = 1
  tikv_node_count            = 2
  tidb_node_count            = 1
  monitor_node_count         = 1
}
```

> **Note:**
>
> The `cluster_name` of each cluster must be unique.

> > **Note:**

Reviewer 3: @yikeke The two notes here need to be combined together, right?

Reviewer 4: Yes, avoiding two consecutive notes improves readability. In this particular case, you can use:

```
> **Note:**
>
> - The `cluster_name` of each cluster must be unique.
>
> - The actual number of nodes created are multiplied by the number of availability zones in the region.
```

>
> The actual number of nodes created are multiplied by the number of availability zones in the region.

You can get the addresses for TiDB and the monitoring service of the created cluster via `kubectl`. If you want the Terraform script to print this information, you can add `output` sections in `outputs.tf`:

```hcl
output "how_to_connect_to_example_tidb_cluster_from_bastion" {
  value = module.example-tidb-cluster.how_to_connect_to_tidb_from_bastion
}
```

This will print out the exact command to use to connect to the TiDB cluster we just created from the bastion instance.

## Scale

> To scale the TiDB cluster, modify `tikv_count` and `tidb_count`, to your desired count, and run `terraform apply`.

Reviewer 3: should be `tikv_count` or `tidb_count`?

terraform destroy
```

> **Note:**
>
> When `terraform destroy` is running, an error with the following message might occur: `Error reading Container Cluster "tidb": Cluster "tidb" has status "RECONCILING" with message""`. This happens when GCP is upgrading the kubernetes master node, which it does automatically at times. While this is happening, it is not possible to delete the cluster. When it is done, run `terraform destroy` again.

### Delete disks after use

> If you no longer need the data and would like to delete the disks that were being used, there are a couple ways to do so:

Reviewer: ```suggestion
If you no longer need the data and would like to delete the disks in use, there are a couple of ways to do so:
```

> - Manual deletion: this can be done either in Google Cloud Console, or with `gcloud`.

Reviewer: ```suggestion
- Manual deletion: do this either in Google Cloud Console or using the `gcloud` command-line tool.
```

> - Setting the kubernetes persistent volume reclaim policy to `Delete` prior to `terraform destroy`: This can be accomplished by running the following `kubectl` command before running `terraform destroy`

Reviewer: ```suggestion
- Setting the Kubernetes persistent volume reclaiming policy to `Delete` prior to executing `terraform destroy`: Do this by running the following `kubectl` command before `terraform destroy`
```

```bash
kubectl --kubeconfig /path/to/kubeconfig/file get pvc -n namespace-of-tidb-cluster -o jsonpath='{.items[*].spec.volumeName}'|fmt -1 | xargs -I {} kubectl --kubeconfig /path/to/kubeconfig/file patch pv {} -p '{"spec":{"persistentVolumeReclaimPolicy":"Delete"}}'
```

> This command will get the persistent volume claims in the TiDB cluster namespace and set the reclaim policy of the persistent volumes to `Delete`. When the PVCs are deleted during `terraform destroy`, the disks will be deleted as well.

Reviewer: ```suggestion
This command will get the persistent volume claims in the TiDB cluster namespace and set the reclaiming policy of the persistent volumes to `Delete`. When the PVCs are deleted during `terraform destroy`, the disks are deleted as well.
```

> There is a script that simplifies this process in `deploy/gcp` relative to the root directory of the repository, called `change-pv-reclaimpolicy.sh`. This script can be run as follows:

Reviewer: ```suggestion
Below is a script called `change-pv-reclaimpolicy.sh`.  It simplifies the above process in `deploy/gcp` relative to the root directory of the repository, . 
```
